addons/account/models/account_account.py

```diff
@@ -1103,61 +1111,58 @@ def _get_query_company_id(model):
                 )
 
         # Step 1: Check access rights.
-        self.check_access('write')
-        if forbidden_companies := (self.sudo().company_ids - self.env.companies):
-            raise UserError(_(
-                "You do not have the right to perform this operation as you do not have access to the following companies: %s.",
-                ", ".join(c.name for c in forbidden_companies)
-            ))
+        self._check_action_unmerge_possible()
 
         # Step 2: Create new accounts.
         base_company = self.env.company if self.env.company in self.company_ids else self.company_ids[0]
-        companies_to_update = self.company_ids - base_company
-        check_company_fields = {fname for fname, field in self._fields.items() if field.relational and field.check_company}
+        self_env = self.with_context({'allowed_company_ids': self.company_ids.ids})
```

Comment:
This is reordering the companies, so `base_company` might not be `self.env.company` anymore

Reply:
I've changed it so that we take `(self.env.companies | self.company_ids).ids`
If I understand correctly how `union` works, that should keep all the active companies first, in the order in which they previously were in `allowed_company_ids`.